src/system/user.cr

```diff
@@ -0,0 +1,159 @@
+require "c/pwd"
+require "c/sys/limits"
+
+struct System::User
+  {% if flag?(:darwin) || flag?(:openbsd) || flag?(:freebsd) %}
+    UID_MAX = LibC::UID_MAX
+  {% elsif flag?(:linux) %}
+    UID_MAX = 0xffffffff_u32
+  {% else %}
+    {{ raise "Unsupported platform, only Darwin, OpenBSD, FreeBSD, and Linux (GNU, musl) are supported." }}
+    #UID_MAX = 0xffff_u32 # POSIX Default
+  {% end %}
+
+  # Returns the user name for the given user ID.
+  # Raises `System::User::NotFoundError` if no user exists.
+  def self.name(uid : Int) : String
+    name?(uid) || raise NotFoundError.new("User #{uid.inspect} was not found")
+  end
+
+  # Returns the user name for the given user ID.
+  # Returns `nil` if no user exists.
+  def self.name?(uid : Int) : String?
+    check_uid_in_bounds(uid)
+    user_struct = LibC.getpwuid(uid)
+    return nil if !user_struct
+    String.new(user_struct.value.pw_name)
+  end
+
+  # Returns the user ID for the given user name.
+  # Raises `System::User::NotFoundError` if no user exists.
+  def self.uid(username : String) : UInt32
+    uid?(username) || raise NotFoundError.new("User #{username.inspect} was not found")
+  end
+
+  # Returns the user ID for the given user name.
+  # Returns `nil` if no user exists.
+  def self.uid?(username : String) : UInt32?
+    username.check_no_null_byte
+    user_struct = LibC.getpwnam(username)
+    return nil if !user_struct
+    user_struct.value.pw_uid
+  end
+
+  # Returns the user specified by the user ID.
+  # Raises `System::User::NotFoundError` if not found.
+  def self.get(uid : Int) : User
+    get?(uid) || raise NotFoundError.new("User #{uid.inspect} was not found")
+  end
+
+  # Returns the user specified by the user ID.
+  # Returns `nil` if not found.
+  def self.get?(uid : Int) : User?
+    check_uid_in_bounds(uid)
+
+    buf_size = LibC.sysconf(LibC::SC_GETPW_R_SIZE_MAX)
+    buf_size = 16384 if buf_size < 0
+
+    pw_store = Pointer(LibC::Passwd).malloc(1)
+    buf = Pointer(UInt8).malloc(buf_size)
+    result = pointerof(pw_store)
+
+    success = LibC.getpwuid_r(uid, pw_store, buf, buf_size, result)
+
+    if result.value.null?
+      return nil if success == 0
+      raise "getpwuid_r failed"
```

Comment:
Please raise Errno, and we can avoid an allocation (LibC::Passwd) and cache `_SC_GETPW_R_SIZE_MAX`:

```crystal
private GETPW_R_SIZE_MAX = begin
  size = LibC.sysconf(LibC::SC_GETPW_R_SIZE_MAX)
  size < 0 ? 16384 : size
end

def self.get?(uid : Int) : User?
  pwd = uninitialized LibC::Passwd
  buf = Bytes.new(GETPW_R_SIZE_MAX)
  result = pointerof(pwd)
 
  unless LibC.getpwuid_r(uid, pointerof(pwd), buf, buf.size, pointerof(result)) == 0
    raise Errno.new("getpwuid_r")
  end

  new(pwd) # or new(result) to pass pwd*
end
```

Reply:
Except I don't account for `result*` being NULL while the function returned `0` which is valid, so yeah:

```crystal
ret = LibC.getpwuid_r(uid, pointerof(pwd), buf, buf.size, pointerof(result))

unless result
  return if ret == 0 # not found
  raise Errno.new("getpwuid_r")
end

new(pwd)

```